File: firmware/core2/src/domain/geo.cpp
#include "domain/geo.h"

#include <cmath>

namespace ofd {

namespace {
constexpr double kEarthRadiusKm = 6371.0088;

double toRadians(double deg) { return deg * M_PI / 180.0; }
double toDegrees(double rad) { return rad * 180.0 / M_PI; }
}  // namespace

double haversineDistanceKm(double lat1, double lon1, double lat2, double lon2) {
  const double dLat = toRadians(lat2 - lat1);
  const double dLon = toRadians(lon2 - lon1);
  const double a = std::sin(dLat / 2) * std::sin(dLat / 2) +
                    std::cos(toRadians(lat1)) * std::cos(toRadians(lat2)) *
                        std::sin(dLon / 2) * std::sin(dLon / 2);
  const double c = 2 * std::atan2(std::sqrt(a), std::sqrt(1 - a));
  return kEarthRadiusKm * c;
}

double initialBearingDeg(double lat1, double lon1, double lat2, double lon2) {
  const double phi1 = toRadians(lat1);
  const double phi2 = toRadians(lat2);
  const double dLon = toRadians(lon2 - lon1);
  const double y = std::sin(dLon) * std::cos(phi2);
  const double x = std::cos(phi1) * std::sin(phi2) - std::sin(phi1) * std::cos(phi2) * std::cos(dLon);
  const double bearing = toDegrees(std::atan2(y, x));
  double normalized = std::fmod(bearing + 360.0, 360.0);
  if (normalized < 0) normalized += 360.0;
  return normalized;
}

bool isWithinCircle(double lat, double lon, double centerLat, double centerLon, double radiusKm) {
  return haversineDistanceKm(lat, lon, centerLat, centerLon) <= radiusKm;
}

}  // namespace ofd

File: services/gateway/src/lib/geo.ts
const EARTH_RADIUS_KM = 6371.0088;

function toRadians(deg: number): number {
  return (deg * Math.PI) / 180;
}

function toDegrees(rad: number): number {
  return (rad * 180) / Math.PI;
}

/** Great-circle distance between two points, in kilometers. */
export function haversineDistanceKm(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number,
): number {
  const dLat = toRadians(lat2 - lat1);
  const dLon = toRadians(lon2 - lon1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRadians(lat1)) * Math.cos(toRadians(lat2)) * Math.sin(dLon / 2) ** 2;
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return EARTH_RADIUS_KM * c;
}

/** Initial bearing from point 1 to point 2, in degrees [0, 360). */
export function initialBearingDeg(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const phi1 = toRadians(lat1);
  const phi2 = toRadians(lat2);
  const dLon = toRadians(lon2 - lon1);
  const y = Math.sin(dLon) * Math.cos(phi2);
  const x = Math.cos(phi1) * Math.sin(phi2) - Math.sin(phi1) * Math.cos(phi2) * Math.cos(dLon);
  const bearing = toDegrees(Math.atan2(y, x));
  return (bearing + 360) % 360;
}

/** True if (lat, lon) is within radiusKm of (centerLat, centerLon). */
export function isWithinCircle(
  lat: number,
  lon: number,
  centerLat: number,
  centerLon: number,
  radiusKm: number,
): boolean {
  return haversineDistanceKm(lat, lon, centerLat, centerLon) <= radiusKm;
}
